File: src/day3.c
#include <stdio.h>
#include <string.h>

#define MAX_LINE_LENGTH     1000

struct Strategy {
    int right, down;
};

long int treeCost(struct Strategy* s, int n, char* line) {

    if (n % (*s).down != 0) return 0;
    int len = strlen(line)- 1;
    int downSteps = n / (*s).down;
    int index = (downSteps * (*s).right) % len;
    if (line[index] == '.') return 0;
    return 1;
}

long int eachLine (struct Strategy* s)
{
    FILE* file = fopen("puzzle-inputs/3.txt", "r");
    char line[MAX_LINE_LENGTH];

    long int total = 0;
    int index = 0;
    while(fgets(line, FILENAME_MAX, file))
    {
        total += treeCost(s, index, line);
        index++;
    }
    return total;
}

int main () {
    struct Strategy strategies[] =
        { { 3, 1 }
        , { 1, 1 }
        , { 5, 1 }
        , { 7, 1 }
        , { 1, 2 }
        };

    long int total = eachLine(&strategies[0]);
    printf("Total trees found in path %ld\n", total);

    long int all = 1;
    for (int n = 5; n--;) {
        printf("For strategy %d: %ld\n", n, eachLine(&strategies[n]));
        all *= eachLine(&strategies[n]);
    }
    printf("Total trees found multiplied %ld\n", all);

}

File: src/day3.rs
use std::fs;
use std::str::Lines;

const tree: char = '#';

pub fn main() {
    let file = fs::read_to_string("puzzle-inputs/3.txt").expect("Unable to read input file");
    let initial_strategy = Strategy { down: 1, right: 3 };

    println!(
        "Trees hit using strategy one: {}",
        count_trees(file.lines(), &initial_strategy)
    );

    let strategies = vec![
        Strategy { down: 1, right: 1 },
        Strategy { down: 1, right: 3 },
        Strategy { down: 1, right: 5 },
        Strategy { down: 1, right: 7 },
        Strategy { down: 2, right: 1 },
    ];

    let multiplied_trees = strategies
        .iter()
        .map(|s| count_trees(file.lines(), &s))
        .fold(1, |acc, v| acc * v);

    println!(
        "Trees hit using multiple strategies multiplied: {}",
        multiplied_trees
    );
}

struct Strategy {
    down: u8,
    right: u8,
}

fn count_trees(lines: Lines, strategy: &Strategy) -> usize {
    lines
        .enumerate()
        .map(|(ix, line)| hits_tree(ix, line, strategy))
        .map(|hit| if hit { 1 } else { 0 })
        .sum()
}

fn hits_tree(ix: usize, line: &str, s: &Strategy) -> bool {
    if (ix % (s.down as usize)) != 0 {
        return false;
    }

    let down_rounds = ix / s.down as usize;
    let right_position = down_rounds * s.right as usize;
    let pass_ix = right_position % line.len();
    line.chars().nth(pass_ix) == Some(tree)
}
